cmd/util/cmd/epochs/cmd/recover.go

```diff
@@ -83,103 +72,142 @@ func generateRecoverEpochTxArgs(cmd *cobra.Command, args []string) {
 		log.Fatal().Err(err).Msg("failed")
 	}
 
-	// extract arguments from recover epoch tx from snapshot
-	txArgs := extractRecoverEpochArgs(snapshot)
+	return snapshot
+}
 
-	// encode to JSON
-	encodedTxArgs, err := epochcmdutil.EncodeArgs(txArgs)
-	if err != nil {
-		log.Fatal().Err(err).Msg("could not encode recover epoch transaction arguments")
-	}
+// generateRecoverEpochTxArgs generates recover epoch transaction arguments from a root protocol state snapshot and writes it to a JSON file
+func generateRecoverEpochTxArgs(getSnapshot func() *inmem.Snapshot) func(cmd *cobra.Command, args []string) {
+	return func(cmd *cobra.Command, args []string) {
+		stdout := cmd.OutOrStdout()
 
-	// write JSON args to stdout
-	_, err = stdout.Write(encodedTxArgs)
-	if err != nil {
-		log.Fatal().Err(err).Msg("could not write jsoncdc encoded arguments")
+		// extract arguments from recover epoch tx from snapshot
+		txArgs := extractRecoverEpochArgs(getSnapshot())
+
+		// encode to JSON
+		encodedTxArgs, err := epochcmdutil.EncodeArgs(txArgs)
+		if err != nil {
+			log.Fatal().Err(err).Msg("could not encode recover epoch transaction arguments")
+		}
+
+		// write JSON args to stdout
+		_, err = stdout.Write(encodedTxArgs)
+		if err != nil {
+			log.Fatal().Err(err).Msg("could not write jsoncdc encoded arguments")
+		}
 	}
 }
 
 // extractResetEpochArgs extracts the required transaction arguments for the `resetEpoch` transaction
 func extractRecoverEpochArgs(snapshot *inmem.Snapshot) []cadence.Value {
 	epoch := snapshot.Epochs().Current()
-	ids, err := epoch.InitialIdentities()
+
+	ids, err := snapshot.Identities(filter.IsValidProtocolParticipant)
 	if err != nil {
-		log.Fatal().Err(err).Msg("failed to get initial identities for current epoch")
+		log.Fatal().Err(err).Msg("failed to get  valid protocol participants from snapshot")
 	}
 
-	currentEpochDKG, err := epoch.DKG()
+	// separate collector nodes by internal and partner nodes
+	collectors := ids.Filter(filter.HasRole[flow.Identity](flow.RoleCollection))
+	internalCollectors := make(flow.IdentityList, 0)
+	partnerCollectors := make(flow.IdentityList, 0)
+
+	log.Info().Msg("collecting internal node network and staking keys")
```

Comment:
~We should confirm `log` is logging to `stderr`. Otherwise we won't have just the JSON output in `stdout` and won't be able to direct JSON output to a file.~

Ah, you're already effectively confirming this in [the tests](https://github.com/onflow/flow-go/pull/5576/files/7d09f47d73fbf7930a38d2e591d783209c7f92a2..3a8ec88c685d70ab7378df7d7eee62a50fcfa4fb#diff-0902a10a5b4940312867be36693a067db3a99c29b5af1d9012d2f654a504c713R53-R56): perfect 👌